### DISCUSSION TITLE
[NEED HELP] mosaik trying to step a simulator before it's max advance time (maybe?)

### DISCUSSION THREAD

**Discussion by talhaibnaziz** [OFFIS-mosaik/mosaik#22] [Q&A] [answered]
## Overview
Hello, I am working on a smart grid co-simulation project which uses `mosaik` to couple 4 simulators. While simulating, I get an error caused by one of the simulators trying to step into the past. I am confused about whether this is a bug in the co-simulator platform or, in one of the simulator configurations.

Question: Is this a bug? If not, how to create the correct behavior?

## Co-simulation configuration
There are 4 simulators as follows:
* Power Flow
  - This is simulated using OpenDSS and simulates the electrical network.
  - It uses the high-level API
  - Marked as a `hybrid` simulator and is connected to the packet network (cyclic data flow)
* Packet network
  - This is simulated using NS3 and uses the low-level API
  - This is marked as an `event-based` simulator and is connected to both the Power Flow and Controller
* Controller
  - Takes the smart grid decisions but is irrelevant in this discussion
  - Marked as `hybrid` and is connected to packet network only.
  - This also uses high-level API
* Collector
  - Connected to all simulators and only collects data from them (no cyclic connections)

In general how the co-simulation works is, the power flow generates sensory data, carried by the packet network to the Controller location (and then passes the data through Mosaik to Controller). The Controller takes a decision, which is carried over by the packet network to the location of the sensors and then passes this info to the Power Flow simulator.

## Debugging info
The error is generated within mosaik and as as follows:
```
`assert time >= self.time, "cannot progress backwards"`
```

I put the following lines in the mosaik python files to help debug this:
```
if world.rt_factor:
	rt_passed = perf_counter() - sim.rt_start
	rt_progress = [TieredTime(ceil(rt_passed / world.rt_factor))]
else:
	rt_progress = []
new_progress = min([
	*pre_sim_induced_progress,
	*next_step_progress,
	*current_step_prog,
	*rt_progress,
	TieredTime(world.until) + sim.from_world_time,
])
if (new_progress < sim.progress.time): #-- New lines added
	breakpoint()                       #-- to debug this error
sim.progress.set(new_progress)
sim.tqdm.update(new_progress.time - sim.tqdm.n)
```

What happens is, the Power Flow network steps to 1200 and then provides the next step time for 1300. After that, the packet network steps to 1201 and provides with the next step time of 1234. From my understanding, what should happen next is Power Flow should step to 1233 first even though it's next step time is 1300 - because it will now be triggered by NS3 to step with max advance time of 1234. However, mosaik throws the assertion error because internally (in mosaik) it steps Power Flow to 1300 first and then tries to step it to 1234 - i.e., it is getting triggered AFTER already stepping to 1300 (even though Power Step was not called yet).

Here are some details from my debugging to find out what was happening:
```
simulator_pflow::step time =  1200  Max Advance =  1201
simulator_pflow::step inputs =  {'Actuator_650.0': {'t': {'PktNetSim-0.Transp_632-650.0': [1001.0]}, 'v': {'PktNetSim-0.Transp_632-650.0': ['-1']}}}
Generating load for:  1120
Generating load for:  1200
simulator_pflow::step Propagation delay = 199.0
simulator_pflow::step next_step =  1300
simulator_pflow::get_data INPUT {'Sensor_611-632.0.0': ['t', 'v', 't', 'v'], 'Actuator_650.0': ['t', 'v'], 'Prober_611.0.2': ['t', 'v'], 'Prober_650.1.2': ['t', 'v'], 'Prober_611.2.0': ['t', 'v'], 'Prober_650.3.2': ['t', 'v']}
simulator_pflow::get_data data: {'Sensor_611-632.0.0': {'v': [np.float64(2171.371664487613)], 't': [1200.000000001]}, 'Actuator_650.0': {'v': ['-1'], 't': [1200]}, 'Prober_611.0.2': {'v': [np.float64(2171.371664487613)], 't': [1200]}, 'Prober_650.1.2': {'v': [0.9812499999999997], 't': [1200]}, 'Prober_611.2.0': {'v': [62.09777245259694], 't': [1200]}, 'Prober_650.3.2': {'v': [0.9999805174946796], 't': [1200]}}
> /home/talha-ibn-aziz/workspace/co_sim_platform/virtualenv/cosimul/lib/python3.12/site-packages/mosaik/scheduler.py(487)advance_progress()
-> sim.progress.set(new_progress)
(Pdb) new_progress
1201
(Pdb) new_progress
1201
(Pdb) sim.progress.time
1200
(Pdb) cont
MosaikSim::step ROLLBACK TIME !! (632->650) NS3 Time: 1127 val_T: 1001
> /home/talha-ibn-aziz/workspace/co_sim_platform/virtualenv/cosimul/lib/python3.12/site-packages/mosaik/scheduler.py(487)advance_progress()
-> sim.progress.set(new_progress)
  3%|███▏                                                                                                      | 1201/40000 [04:20<43:27, 14.88stesim.progresnew_progress
1202
(Pdb) sim.progress.time
1201
(Pdb) cont
> /home/talha-ibn-aziz/workspace/co_sim_platform/virtualenv/cosimul/lib/python3.12/site-packages/mosaik/scheduler.py(487)advance_progress()
-> sim.progress.set(new_progress)
(Pdb) new_progress
1202
(Pdb) sim.progress.time
1202
(Pdb) cont
  3%|███▏                                                                                                    | 1202/40000 [04:55<8:08:10,  1.32steps/s]> /home/talha-ibn-aziz/workspace/co_sim_platform/virtualenv/cosimul/lib/python3.12/site-packages/mosaik/scheduler.py(487)advance_progress()
-> sim.progress.set(new_progress)
(Pdb) new_progress
1300
(Pdb) sim.progress.time
1202
(Pdb) cont
  3%|███▎                                                                                                    | 1276/40000 [05:44<7:33:14,  1.42steps/s]> /home/talha-ibn-aziz/workspace/co_sim_platform/virtualenv/cosimul/lib/python3.12/site-packages/mosaik/scheduler.py(487)advance_progress()
-> sim.progress.set(new_progress)
(Pdb) sim.progress.time
1300
(Pdb) new_progress
1234
(Pdb) 
```

**Comment by eike** [comment, marked as answer]
This looks like a bug! Thank you for reporting it, and for the initial investigation. I have also run into this issue myself recently, here's an issue to track it: https://gitlab.com/mosaik/mosaik/-/issues/246.

I will look into it, but it might take a couple of weeks, as there are other deadlines looming. In my case, I was able to resolve the error by adding `async_requests=True` to some connections. (This makes mosaik a bit more conservative in advancing the progress.) This is a workaround, *not* the proper solution, which is to fix mosaik.

**Comment by talhaibnaziz** [comment]
Hello, thanks for the update - I get the cyclic connections error when I use `async_requests=True`. And, I also see that this flag is now deprecated in the documentation - so probably this does not work anymore as the cycle resolver in mosaik?

**Comment by talhaibnaziz** [comment]
Thanks for the advice, I resolved it by downgrading mosaik from 3.33 to 3.0 and using `weak=True` to resolve the cyclic dependencies.
